.github/workflows/build.yml

```diff
@@ -36,15 +36,23 @@ jobs:
       
     - name: Install Toolchains
       run: |
+        # ARM & RISC-V GCC from xpack
         npm install --global xpm
         xpm install --global @xpack-dev-tools/arm-none-eabi-gcc@latest
         xpm install --global @xpack-dev-tools/riscv-none-embed-gcc@latest
-        wget http://software-dl.ti.com/msp430/msp430_public_sw/mcu/msp430/MSPGCC/8_3_0_0/exports/msp430-gcc-8.3.0.16_linux64.tar.bz2 -O /tmp/msp430-gcc.tar.bz2
-        tar -C $HOME -xaf /tmp/msp430-gcc.tar.bz2
         echo "::add-path::`echo $HOME/opt/xPacks/@xpack-dev-tools/arm-none-eabi-gcc/*/.content/bin`"
         echo "::add-path::`echo $HOME/opt/xPacks/@xpack-dev-tools/riscv-none-embed-gcc/*/.content/bin`"
+
+        # TI MSP430 GCC
+        wget http://software-dl.ti.com/msp430/msp430_public_sw/mcu/msp430/MSPGCC/8_3_0_0/exports/msp430-gcc-8.3.0.16_linux64.tar.bz2 -O /tmp/msp430-gcc.tar.bz2
+        tar -C $HOME -xaf /tmp/msp430-gcc.tar.bz2
         echo "::add-path::`echo $HOME/msp430-gcc-*_linux64/bin`"
-      
+
+        # ESP IDF
+        git clone https://github.com/espressif/esp-idf.git $HOME/esp-idf
```

Comment:
To make the download smaller, perhaps do a shallow clone `git --depth 1 clone $(URL) `. In the future, some particular release should be tagged. As a side note, it looks like they have one branch (4.1) that supports the beta hardware, and v4.2 will support production.

Reply:
good idea, though this is purely for CI only. Since the repo currently doesn't include ESP-IDF as submodules since it is not simply mcu driver like others. And people probably has their own ESP-IDF setup somewhere already.

Reply:
Agreed; I feel bad for the CI servers doing extra work.

Reply:
I will apply your suggestion in the next push, Faster CI is what we want as well :)

Reply:
Related to that, it looks like we can also use the `action/cache` module to cache copies of the MSP430 downloads. I don't know if that would quicken things or not.

Reply:
Yeah, with Travis, cache is needed for downloaded file like msp430 gcc. Though for action i it is super fast, I guess it does have some caching scheme in the background

Reply:
By the way, [there is a docker image](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/tools/idf-docker-image.html) of ESP-IDF which you could use for CI: `espressif/idf`. Typically downloading that image in CI is faster than cloning IDF plus installing the tools.

Reply:
docker is a great option, though, I am not familiar with docker, and spent too much time tweaking CI. I separated esp32s2 from other builds matrix and it build much quicker now. Happy with the result for the time being. We will visit the docker later on :)